### PR TITLE
Upgrade to OpenAI v4 (extends #2353)

### PR DESCRIPTION
**Pull request by jacoblee93**
Extends #2353 with some backcompat shims

Wasn't able to push on top of the branch @nsbradford - can you have a look?

CC @rattrayalex as well

### CODE REVIEW

File: langchain/src/chat_models/openai.ts [hunk 1]
```diff
@@ -418,58 +408,9 @@ export class ChatOpenAI
...
-        if (err) {
-          throw err;
-        }
-        return chunk;
-      },
-      [Symbol.asyncIterator]() {
-        return this;
-      },
-    };
+    if (options.signal?.aborted) {
```
Comment (jacoblee93) [inline langchain/src/chat_models/openai.ts:411]: CC @rattrayalex if the user aborts a stream, the SDK no longer throws an error. This is different from `fetch`'s behavior

File: langchain/src/util/async_caller.ts [hunk 2]
```diff
@@ -80,7 +80,9 @@ export class AsyncCaller {
               if (
                 error.message.startsWith("Cancel") ||
                 error.message.startsWith("TimeoutError") ||
-                error.message.startsWith("AbortError")
+                error.message.startsWith("AbortError") ||
+                error.message.includes("Request was aborted") ||
```
Comment (jacoblee93) [inline langchain/src/util/async_caller.ts:84]: @rattrayalex is there any way to detect these errors other than a constructor check? This will fail in minified environments...

File: langchain/src/chat_models/openai.ts [hunk 3]
```diff
@@ -712,56 +589,83 @@ export class ChatOpenAI
...
-        baseOptions: {
-          timeout: this.timeout,
-          ...this.clientConfig.baseOptions,
-        },
-      });
+        baseURL: endpoint,
+        timeout: this.timeout,
+        maxRetries: 0,
+      };
+      if (!params.baseURL) {
```
Comment (jacoblee93) [inline langchain/src/chat_models/openai.ts:648]: @rattrayalex would be nice to not have to do this `delete` - could `baseURL: undefined` default to just the default OAI setting?

**Comment by jacoblee93** [comment]
Some more work to do, perhaps with @rattrayalex's input?

File: langchain/src/chat_models/openai.ts (hunk 1 above)
Comment (rattrayalex) [reply langchain/src/chat_models/openai.ts:411]: Yes, that's intentional – other than backcompat, does that cause problems / is there a reason it's not desirable behavior?

BTW, you likely want to define an `AbortError` class and throw it below – people identify these by checking that the `.name` of the error is `AbortError`.

File: langchain/src/chat_models/openai.ts (hunk 3 above)
Comment (rattrayalex) [reply langchain/src/chat_models/openai.ts:648]: Ah, sorry about this - agree that's a bug, I've filed a ticket to fix.

File: langchain/src/util/async_caller.ts (hunk 2 above)
Comment (rattrayalex) [reply langchain/src/util/async_caller.ts:84]: I think you want `error.name === 'AbortError'` ([docs](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal))

Comment (rattrayalex) [reply langchain/src/util/async_caller.ts:84]: Depending on what you want, we also export classes like `APIUserAbortError` and `APIConnectionTimeoutError`; you could check the `name` there too if you don't wan't to do an `instanceof`

Comment (jacoblee93) [reply langchain/src/util/async_caller.ts:84]: `error.name` I think just returned `Error` but will double check.

File: langchain/src/chat_models/openai.ts (hunk 1 above)
Comment (jacoblee93) [reply langchain/src/chat_models/openai.ts:411]: No, we can handle it this way - convention across other LLMs is to throw because they're fetch based so would like to keep it standard for us but no worries on your end then.

File: langchain/src/util/async_caller.ts (hunk 2 above)
Comment (rattrayalex) [reply langchain/src/util/async_caller.ts:84]: got it, that might be because of this line: https://github.com/hwchase17/langchainjs/pull/2389/files/a0596c79e38da6f527a0a4578ba66bb53cfab999#diff-4d0f479be12d9b217277a26941bcfe72cfb9bc6b9bb9209099d6dd3294248bfbR412

which as I noted you may want to do something different with…

File: langchain/src/chat_models/anthropic.ts [hunk 4]
```diff
@@ -377,6 +377,7 @@ export class ChatAnthropic extends BaseChatModel implements AnthropicInput {
         ...this.clientOptions,
         ...options,
         apiKey: this.anthropicApiKey,
+        maxRetries: 0,
```
Comment (nsbradford) [inline langchain/src/chat_models/anthropic.ts:380]: this looks unrelated - was changing the Anthropic client intentional?

Comment (jacoblee93) [reply langchain/src/chat_models/anthropic.ts:380]: Yes, oversight earlier but we handle retries natively through the `AsyncCaller` class for standardization.

**Comment by nsbradford** [comment]
> Some more work to do, perhaps with @rattrayalex's input?

hi @jacoblee93 @rattrayalex are you still going through review? Any assistance you need?

**Comment by jacoblee93** [comment]
Yes, have been quite busy but still plan to merge this after extensive testing!

File: langchain/src/chat_models/openai.ts [hunk 5]
```diff
@@ -323,6 +314,14 @@ export class ChatOpenAI
     this.clientConfig = {
       apiKey: this.openAIApiKey,
       organization: this.organization,
+      baseURL: configuration?.basePath ?? fields?.configuration?.basePath,
+      dangerouslyAllowBrowser: true,
+      defaultHeaders:
```
Comment (jacoblee93) [inline langchain/src/chat_models/openai.ts:319]: Am planning to just support previously passed `basePath`, `headers`, and `params` but could shim more things in here.

File: langchain/src/chat_models/tests/chatopenai.int.test.ts [hunk 6]
```diff
@@ -63,7 +63,7 @@ test("Test ChatOpenAI Generate throws when one of the calls fails", async () =>
     chat.generate([[message], [message]], {
       signal: AbortSignal.timeout(10),
     })
-  ).rejects.toThrow("Cancel: canceled");
+  ).rejects.toThrow();
```
Comment (jacoblee93) [inline langchain/src/chat_models/tests/chatopenai.int.test.ts:66]: Technically breaking but 🤷

File: langchain/src/chat_models/tests/chatopenai.int.test.ts [hunk 7]
```diff
@@ -242,7 +242,7 @@ test("Test OpenAI with timeout in call options and node adapter", async () => {
   const model = new ChatOpenAI({ maxTokens: 5 });
   await expect(() =>
     model.call([new HumanMessage("Print hello world")], {
-      options: { timeout: 10, adapter: undefined },
+      options: { timeout: 10 },
```
Comment (jacoblee93) [inline langchain/src/chat_models/tests/chatopenai.int.test.ts:245]: Custom adapters will no longer work but it's probably better long-term anyway

**Comment by jacoblee93** [comment]
@rattrayalex @nsbradford @nfcampos it's massive but would appreciate a last look here if you have time!

File: langchain/src/llms/openai-chat.ts [hunk 8]
```diff
@@ -531,7 +512,8 @@ export class PromptLayerOpenAIChat extends OpenAIChat {
         const promptLayerRespBody = await promptLayerTrackRequest(
           this.caller,
           "langchain.PromptLayerOpenAIChat",
-          { ...this._identifyingParams(), prompt },
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          { ...this._identifyingParams(), prompt } as any,
```
Comment (nsbradford) [inline langchain/src/llms/openai-chat.ts:516]: is this `any` cast necessary?

**Review by nsbradford — APPROVED** [review]
lgtm

**Comment by jacoblee93** [comment]
> lgtm

Thank you!

**Comment by jacoblee93** [comment]
No turning back now! Thanks all!

**Comment by siisee11** [comment]
Can you briefly explain how this PR can solve https://github.com/hwchase17/langchainjs/issues/2346?
